File: wcslib/C/test/tspx.c
/*============================================================================

  WCSLIB 4.7 - an implementation of the FITS WCS standard.
  Copyright (C) 1995-2011, Mark Calabretta

  This file is part of WCSLIB.

  WCSLIB is free software: you can redistribute it and/or modify it under the
  terms of the GNU Lesser General Public License as published by the Free
  Software Foundation, either version 3 of the License, or (at your option)
  any later version.

  WCSLIB is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
  more details.

  You should have received a copy of the GNU Lesser General Public License
  along with WCSLIB.  If not, see <http://www.gnu.org/licenses/>.

  Correspondence concerning WCSLIB may be directed to:
    Internet email: mcalabre@atnf.csiro.au
    Postal address: Dr. Mark Calabretta
                    Australia Telescope National Facility, CSIRO
                    PO Box 76
                    Epping NSW 1710
                    AUSTRALIA

  Author: Mark Calabretta, Australia Telescope National Facility
  http://www.atnf.csiro.au/~mcalabre/index.html
  $Id: tspx.c,v 4.7 2011/02/07 07:03:42 cal103 Exp $
*=============================================================================
*
* tspx tests the spectral transformation routines for closure.
*
*---------------------------------------------------------------------------*/

#include <math.h>
#include <stdio.h>

#include <spx.h>


#define NSPEC 9991

const double tol = 1.0e-9;

void closure (const char *, const char *, double, int (*)(), int (*)(),
              const double [], double []);

const double C = 2.99792458e8;


int main()

{
  int    stat[NSPEC], status;
  double restfrq, restwav, step;
  double awav[NSPEC], freq[NSPEC], spc1[NSPEC], spc2[NSPEC], velo[NSPEC],
         wave[NSPEC];
  struct spxprm spx;

  register int j, k;


  printf(
    "Testing closure of WCSLIB spectral transformation routines (tspx.c)\n"
    "-------------------------------------------------------------------\n");

  /* List status return messages. */
  printf("\nList of spx status return values:\n");
  for (status = 1; status <= 4; status++) {
    printf("%4d: %s.\n", status, spx_errmsg[status]);
  }

  restfrq = 1420.40595e6;
  restwav = C/restfrq;


  /* Exercise specx(). */
  printf("\nTesting spectral cross-conversions (specx).\n\n");
  if ((status = specx("VELO", 4.3e5, restfrq, restwav, &spx))) {
    printf("specx ERROR %d: %s.\n", status, spx_errmsg[status]);
    return 1;
  }

  printf("    restfrq:%20.12E\n", spx.restfrq);
  printf("    restwav:%20.12E\n", spx.restwav);
  printf("   wavetype:%3d\n",     spx.wavetype);
  printf("   velotype:%3d\n",     spx.velotype);
  printf("\n");
  printf("       freq:%20.12E\n", spx.freq);
  printf("       afrq:%20.12E\n", spx.afrq);
  printf("       ener:%20.12E\n", spx.ener);
  printf("       wavn:%20.12E\n", spx.wavn);
  printf("       vrad:%20.12E\n", spx.vrad);
  printf("       wave:%20.12E\n", spx.wave);
  printf("       vopt:%20.12E\n", spx.vopt);
  printf("       zopt:%20.12E\n", spx.zopt);
  printf("       awav:%20.12E\n", spx.awav);
  printf("       velo:%20.12E\n", spx.velo);
  printf("       beta:%20.12E\n", spx.beta);
  printf("\n");

  printf("dfreq/dafrq:%20.12E\n", spx.dfreqafrq);
  printf("dafrq/dfreq:%20.12E\n", spx.dafrqfreq);

  printf("dfreq/dener:%20.12E\n", spx.dfreqener);
  printf("dener/dfreq:%20.12E\n", spx.denerfreq);

  printf("dfreq/dwavn:%20.12E\n", spx.dfreqwavn);
  printf("dwavn/dfreq:%20.12E\n", spx.dwavnfreq);

  printf("dfreq/dvrad:%20.12E\n", spx.dfreqvrad);
  printf("dvrad/dfreq:%20.12E\n", spx.dvradfreq);

  printf("dfreq/dwave:%20.12E\n", spx.dfreqwave);
  printf("dwave/dfreq:%20.12E\n", spx.dwavefreq);

  printf("dfreq/dawav:%20.12E\n", spx.dfreqawav);
  printf("dawav/dfreq:%20.12E\n", spx.dawavfreq);

  printf("dfreq/dvelo:%20.12E\n", spx.dfreqvelo);
  printf("dvelo/dfreq:%20.12E\n", spx.dvelofreq);

  printf("dwave/dvopt:%20.12E\n", spx.dwavevopt);
  printf("dvopt/dwave:%20.12E\n", spx.dvoptwave);

  printf("dwave/dzopt:%20.12E\n", spx.dwavezopt);
  printf("dzopt/dwave:%20.12E\n", spx.dzoptwave);

  printf("dwave/dawav:%20.12E\n", spx.dwaveawav);
  printf("dawav/dwave:%20.12E\n", spx.dawavwave);

  printf("dwave/dvelo:%20.12E\n", spx.dwavevelo);
  printf("dvelo/dwave:%20.12E\n", spx.dvelowave);

  printf("dawav/dvelo:%20.12E\n", spx.dawavvelo);
  printf("dvelo/dawav:%20.12E\n", spx.dveloawav);

  printf("dvelo/dbeta:%20.12E\n", spx.dvelobeta);
  printf("dbeta/dvelo:%20.12E\n", spx.dbetavelo);
  printf("\n");


  /* Construct a linear velocity spectrum. */
  step = (2.0*C/NSPEC) / 2.0;
  for (j = 0, k = -NSPEC; j < NSPEC; j++, k += 2) {
    velo[j] = (k+1)*step;
  }
  printf("\nVelocity range: %.3f to %.3f km/s, step: %.3f km/s\n",
         velo[0]*1e-3, velo[NSPEC-1]*1e-3, (velo[1] - velo[0])*1e-3);

  /* Convert it to frequency. */
  velofreq(restfrq, NSPEC, 1, 1, velo, freq, stat);

  /* Test closure of all two-way combinations. */
  closure("freq", "afrq", 0.0,     freqafrq, afrqfreq, freq, spc1);
  closure("afrq", "freq", 0.0,     afrqfreq, freqafrq, spc1, spc2);

  closure("freq", "ener", 0.0,     freqener, enerfreq, freq, spc1);
  closure("ener", "freq", 0.0,     enerfreq, freqener, spc1, spc2);

  closure("freq", "wavn", 0.0,     freqwavn, wavnfreq, freq, spc1);
  closure("wavn", "freq", 0.0,     wavnfreq, freqwavn, spc1, spc2);

  closure("freq", "vrad", restfrq, freqvrad, vradfreq, freq, spc1);
  closure("vrad", "freq", restfrq, vradfreq, freqvrad, spc1, spc2);

  closure("freq", "wave", 0.0,     freqwave, wavefreq, freq, wave);
  closure("wave", "freq", 0.0,     wavefreq, freqwave, wave, spc2);

  closure("freq", "awav", 0.0,     freqawav, awavfreq, freq, awav);
  closure("awav", "freq", 0.0,     awavfreq, freqawav, awav, spc2);

  closure("freq", "velo", restfrq, freqvelo, velofreq, freq, velo);
  closure("velo", "freq", restfrq, velofreq, freqvelo, velo, spc2);

  closure("wave", "vopt", restwav, wavevopt, voptwave, wave, spc1);
  closure("vopt", "wave", restwav, voptwave, wavevopt, spc1, spc2);

  closure("wave", "zopt", restwav, wavezopt, zoptwave, wave, spc1);
  closure("zopt", "wave", restwav, zoptwave, wavezopt, spc1, spc2);

  closure("wave", "awav", 0.0,     waveawav, awavwave, wave, spc1);
  closure("awav", "wave", 0.0,     awavwave, waveawav, spc1, spc2);

  closure("wave", "velo", restwav, wavevelo, velowave, wave, spc1);
  closure("velo", "wave", restwav, velowave, wavevelo, spc1, spc2);

  closure("awav", "velo", restwav, awavvelo, veloawav, awav, spc1);
  closure("velo", "awav", restwav, veloawav, awavvelo, spc1, spc2);

  closure("velo", "beta", 0.0,     velobeta, betavelo, velo, spc1);
  closure("beta", "velo", 0.0,     betavelo, velobeta, spc1, spc2);

  return 0;
}

/*--------------------------------------------------------------------------*/

void closure (from, to, parm, fwd, rev, spec1, spec2)

const char *from, *to;
double parm;
int (*fwd)(SPX_ARGS);
int (*rev)(SPX_ARGS);
const double spec1[];
double spec2[];

{
  static char skip = '\0';
  int stat1[NSPEC], stat2[NSPEC], status;
  register int j;
  double clos[NSPEC], resid, residmax;

  /* Convert the first to the second. */
  if ((status = fwd(parm, NSPEC, 1, 1, spec1, spec2, stat1))) {
    printf("%s%s ERROR %d: %s.\n", from, to, status, spx_errmsg[status]);
  }

  /* Convert the second back to the first. */
  if ((status = rev(parm, NSPEC, 1, 1, spec2, clos, stat2))) {
    printf("%s%s ERROR %d: %s.\n", to, from, status, spx_errmsg[status]);
  }

  residmax = 0.0;

  /* Test closure. */
  for (j = 0; j < NSPEC; j++) {
    if (stat1[j]) {
      printf("%c%s%s: %s = %.12E -> %s = ???, stat = %d\n", skip, from, to,
             from, spec1[j], to, stat1[j]);
      skip = '\0';
      continue;
    }

    if (stat2[j]) {
      printf("%c%s%s: %s = %.12E -> %s = %.12E -> %s = ???, stat = %d\n",
             skip, to, from, from, spec1[j], to, spec2[j], from, stat2[j]);
      skip = '\0';
      continue;
    }

    if (spec1[j] == 0.0) {
      resid = fabs(clos[j] - spec1[j]);
    } else {
      resid = fabs((clos[j] - spec1[j])/spec1[j]);
      if (resid > residmax) residmax = resid;
    }

    if (resid > tol) {
      printf("%c%s%s: %s = %.12E -> %s = %.12E ->\n          %s = %.12E,  "
             "resid = %.12E\n", skip, from, to, from, spec1[j], to,
             spec2[j], from, clos[j], resid);
      skip = '\0';
    }
  }

  printf("%s%s: Maximum closure residual = %.12E\n", from, to, residmax);
  if (residmax > tol) {
    printf("\n");
    skip = '\0';
  } else {
    skip = '\n';
  }

  return;
}

File: swigtest/wcslib/Fortran/test/tspx.f
*=======================================================================
*
* WCSLIB 4.7 - an implementation of the FITS WCS standard.
* Copyright (C) 1995-2011, Mark Calabretta
*
* This file is part of WCSLIB.
*
* WCSLIB is free software: you can redistribute it and/or modify it
* under the terms of the GNU Lesser General Public License as published
* by the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* WCSLIB is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
* License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with WCSLIB.  If not, see http://www.gnu.org/licenses.
*
* Correspondence concerning WCSLIB may be directed to:
*   Internet email: mcalabre@atnf.csiro.au
*   Postal address: Dr. Mark Calabretta
*                   Australia Telescope National Facility, CSIRO
*                   PO Box 76
*                   Epping NSW 1710
*                   AUSTRALIA
*
* Author: Mark Calabretta, Australia Telescope National Facility
* http://www.atnf.csiro.au/~mcalabre/index.html
* $Id: tspx.f,v 4.7 2011/02/07 07:03:43 cal103 Exp $
*=======================================================================

      PROGRAM TSPX
*-----------------------------------------------------------------------
*
* TSPEC tests the spectral transformation routines for closure.
*
*-----------------------------------------------------------------------
*     Length of spectral axis - see CLOSURE.
      INTEGER   NSPEC
      PARAMETER (NSPEC = 9991)

      INTEGER   J, K, STAT(NSPEC), STATUS
      DOUBLE PRECISION AWAV(NSPEC), C, FREQ(NSPEC), RESTFRQ, RESTWAV,
     :          SPC1(NSPEC), SPC2(NSPEC), STEP, VELO(NSPEC), WAVE(NSPEC)

      INCLUDE 'spx.inc'

      DATA C /2.99792458D8/
*-----------------------------------------------------------------------
      WRITE (*, 10)
 10   FORMAT ('Testing closure of WCSLIB spectral transformation ',
     :        'routines (tspx.f)',/,
     :        '--------------------------------------------------',
     :        '-----------------')

      RESTFRQ = 1420.40595D6
      RESTWAV = C/RESTFRQ

*     Exercise SPECX.
      WRITE (*, 20)
 20   FORMAT (/,'Testing spectral cross-conversions (specx).',/)
      STATUS = SPECX('VELO', 4.3D5, RESTFRQ, RESTWAV, SPX)

      WRITE (*, 30) SPX(1), SPX(2), SPXI(1), SPXI(2), (SPX(J), J=4,14)
 30   FORMAT ('    restfrq:',1PE20.12,/,'    restwav:',1PE20.12,/,
     :        '   wavetype:',I3,/,      '   velotype:',I3,//,
     :        '       freq:',1PE20.12,/,'       afrq:',1PE20.12,/,
     :        '       ener:',1PE20.12,/,'       wavn:',1PE20.12,/,
     :        '       vrad:',1PE20.12,/,'       wave:',1PE20.12,/,
     :        '       vopt:',1PE20.12,/,'       zopt:',1PE20.12,/,
     :        '       awav:',1PE20.12,/,'       velo:',1PE20.12,/,
     :        '       beta:',1PE20.12,/)

      WRITE (*, 40) (SPX(J), J=15,40)
 40   FORMAT ('dfreq/dafrq:',1PE20.12,/,'dafrq/dfreq:',1PE20.12,/,
     :        'dfreq/dener:',1PE20.12,/,'dener/dfreq:',1PE20.12,/,
     :        'dfreq/dwavn:',1PE20.12,/,'dwavn/dfreq:',1PE20.12,/,
     :        'dfreq/dvrad:',1PE20.12,/,'dvrad/dfreq:',1PE20.12,/,
     :        'dfreq/dwave:',1PE20.12,/,'dwave/dfreq:',1PE20.12,/,
     :        'dfreq/dawav:',1PE20.12,/,'dawav/dfreq:',1PE20.12,/,
     :        'dfreq/dvelo:',1PE20.12,/,'dvelo/dfreq:',1PE20.12,/,
     :        'dwave/dvopt:',1PE20.12,/,'dvopt/dwave:',1PE20.12,/,
     :        'dwave/dzopt:',1PE20.12,/,'dzopt/dwave:',1PE20.12,/,
     :        'dwave/dawav:',1PE20.12,/,'dawav/dwave:',1PE20.12,/,
     :        'dwave/dvelo:',1PE20.12,/,'dvelo/dwave:',1PE20.12,/,
     :        'dawav/dvelo:',1PE20.12,/,'dvelo/dawav:',1PE20.12,/,
     :        'dvelo/dbeta:',1PE20.12,/,'dbeta/dvelo:',1PE20.12,/)

*     Construct a linear velocity spectrum.
      STEP = (2D0*C/NSPEC) / 2D0
      K = -NSPEC
      DO 50 J = 1, NSPEC
        VELO(J) = (K+1)*STEP
        K = K + 2
 50   CONTINUE

      WRITE (*, 60) VELO(1)*1D-3, VELO(NSPEC)*1D-3,
     :             (VELO(2) - VELO(1))*1D-3
 60   FORMAT (/,'Velocity range:',F12.3,' to',F11.3,' km/s, step:',
     :        F7.3,' km/s')

*     Convert it to frequency.
      STATUS = VELOFREQ(RESTFRQ, NSPEC, 1, 1, VELO, FREQ, STAT)

*     Test closure of all two-way combinations.
      CALL CLOSURE ('freq', 'afrq', 0D0,     FREQAFRQ, AFRQFREQ, FREQ,
     :              SPC1)
      CALL CLOSURE ('afrq', 'freq', 0D0,     AFRQFREQ, FREQAFRQ, SPC1,
     :              SPC2)

      CALL CLOSURE ('freq', 'ener', 0D0,     FREQENER, ENERFREQ, FREQ,
     :              SPC1)
      CALL CLOSURE ('ener', 'freq', 0D0,     ENERFREQ, FREQENER, SPC1,
     :              SPC2)

      CALL CLOSURE ('freq', 'wavn', 0D0,     FREQWAVN, WAVNFREQ, FREQ,
     :              SPC1)
      CALL CLOSURE ('wavn', 'freq', 0D0,     WAVNFREQ, FREQWAVN, SPC1,
     :              SPC2)

      CALL CLOSURE ('freq', 'vrad', RESTFRQ, FREQVRAD, VRADFREQ, FREQ,
     :              SPC1)
      CALL CLOSURE ('vrad', 'freq', RESTFRQ, VRADFREQ, FREQVRAD, SPC1,
     :              SPC2)

      CALL CLOSURE ('freq', 'wave', 0D0,     FREQWAVE, WAVEFREQ, FREQ,
     :              WAVE)
      CALL CLOSURE ('wave', 'freq', 0D0,     WAVEFREQ, FREQWAVE, WAVE,
     :              SPC2)

      CALL CLOSURE ('freq', 'awav', 0D0,     FREQAWAV, AWAVFREQ, FREQ,
     :              AWAV)
      CALL CLOSURE ('awav', 'freq', 0D0,     AWAVFREQ, FREQAWAV, AWAV,
     :              SPC2)

      CALL CLOSURE ('freq', 'velo', RESTFRQ, FREQVELO, VELOFREQ, FREQ,
     :              VELO)
      CALL CLOSURE ('velo', 'freq', RESTFRQ, VELOFREQ, FREQVELO, VELO,
     :              SPC2)

      CALL CLOSURE ('wave', 'vopt', RESTWAV, WAVEVOPT, VOPTWAVE, WAVE,
     :              SPC1)
      CALL CLOSURE ('vopt', 'wave', RESTWAV, VOPTWAVE, WAVEVOPT, SPC1,
     :              SPC2)

      CALL CLOSURE ('wave', 'zopt', RESTWAV, WAVEZOPT, ZOPTWAVE, WAVE,
     :              SPC1)
      CALL CLOSURE ('zopt', 'wave', RESTWAV, ZOPTWAVE, WAVEZOPT, SPC1,
     :              SPC2)

      CALL CLOSURE ('wave', 'awav', 0D0,     WAVEAWAV, AWAVWAVE, WAVE,
     :              SPC1)
      CALL CLOSURE ('awav', 'wave', 0D0,     AWAVWAVE, WAVEAWAV, SPC1,
     :              SPC2)

      CALL CLOSURE ('wave', 'velo', RESTWAV, WAVEVELO, VELOWAVE, WAVE,
     :              SPC1)
      CALL CLOSURE ('velo', 'wave', RESTWAV, VELOWAVE, WAVEVELO, SPC1,
     :              SPC2)

      CALL CLOSURE ('awav', 'velo', RESTWAV, AWAVVELO, VELOAWAV, AWAV,
     :              SPC1)
      CALL CLOSURE ('velo', 'awav', RESTWAV, VELOAWAV, AWAVVELO, SPC1,
     :              SPC2)

      CALL CLOSURE ('velo', 'beta', 0D0,     VELOBETA, BETAVELO, VELO,
     :              SPC1)
      CALL CLOSURE ('beta', 'velo', 0D0,     BETAVELO, VELOBETA, SPC1,
     :              SPC2)


      END

*=======================================================================

      SUBROUTINE CLOSURE (FROM, TO, PARM, FWD, REV, SPEC1, SPEC2)

      INTEGER   NSPEC
      PARAMETER (NSPEC = 9991)

      LOGICAL   SKIP
      INTEGER   J, STAT1(NSPEC), STAT2(NSPEC), STATUS
      DOUBLE PRECISION CLOS(NSPEC), PARM, RESID, RESIDMAX, SPEC1(NSPEC),
     :          SPEC2(NSPEC), TOL
      CHARACTER FROM*(*), TO*(*)

      PARAMETER (TOL = 1D-9)

      INTEGER  FWD, REV
      EXTERNAL FWD, REV

      SAVE SKIP
      DATA SKIP/.FALSE./
*-----------------------------------------------------------------------

*     Convert the first to the second.
      STATUS = FWD(PARM, NSPEC, 1, 1, SPEC1, SPEC2, STAT1)
      IF (STATUS.NE.0) THEN
        WRITE (*, 10) FROM, TO, STATUS
 10     FORMAT (A,A,' ERROR',I2,'.')
      END IF

*     Convert the second back to the first.
      STATUS = REV(PARM, NSPEC, 1, 1, SPEC2, CLOS, STAT2)
      IF (STATUS.NE.0) THEN
        WRITE (*, 10) TO, FROM, STATUS
      END IF

      RESIDMAX = 0.0

*     Test closure.
      DO 50 J = 1, NSPEC
        IF (STAT1(J).NE.0) THEN
          IF (SKIP) WRITE (*, *)
          WRITE (*, 20) FROM, TO, FROM, SPEC1(J), TO, STAT1(J)
 20       FORMAT (A,A,': ',A,' =',1PE19.12,' -> ',A,
     :            ' = ???, stat = ',I2)
          SKIP = .FALSE.
          GO TO 50
        END IF

        IF (STAT2(J).NE.0) THEN
          IF (SKIP) WRITE (*, *)
          WRITE (*, 30) TO, FROM, FROM, SPEC1(J), TO, SPEC2(J), FROM,
     :                  STAT2(J)
 30       FORMAT (A,A,': ',A,' =',1PE19.12,' -> ',A,' =',1PE19.12,
     :            ' -> ',A,' = ???, stat = ',I2)
          SKIP = .FALSE.
          GO TO 50
        END IF

        IF (SPEC1(J).EQ.0.0) THEN
          RESID = ABS(CLOS(J) - SPEC1(J))
        ELSE
          RESID = ABS((CLOS(J) - SPEC1(J))/SPEC1(J))
          IF (RESID.GT.RESIDMAX) RESIDMAX = RESID
        END IF

        IF (RESID.GT.TOL) THEN
          IF (SKIP) WRITE (*, *)
          WRITE (*, 40) FROM, TO, FROM, SPEC1(J), TO, SPEC2(J), FROM,
     :                  CLOS(J), RESID
 40       FORMAT (A,A,': ',A,' =',1PE19.12,' -> ',A,' =',1PE19.12,
     :            ' ->',/,'          ',A,' =',1PE19.12,',  resid =',
     :            1PE19.12)
          SKIP = .FALSE.
        END IF
 50   CONTINUE

      WRITE (*, 60) FROM, TO, RESIDMAX
 60   FORMAT (A,A,': Maximum closure residual =',1PE19.12)
      IF (RESIDMAX.GT.TOL) THEN
        WRITE (*, *)
        SKIP = .FALSE.
      ELSE
        SKIP = .TRUE.
      END IF

      RETURN
      END
